core/src/Kokkos_CopyViews.hpp

```diff
@@ -303,7 +303,11 @@ struct ViewCopy<ViewTypeA, ViewTypeB, Layout, ExecSpace, 1, iType> {
 
   KOKKOS_INLINE_FUNCTION
   void operator()(const iType& i0) const {
-    a(i0) = static_cast<value_type>(b(i0));
+    if constexpr (std::is_assignable_v<decltype(a(i0)), decltype(b(i0))> &&
+                  !std::is_arithmetic_v<value_type>)
+      a(i0) = b(i0);
+    else
+      a(i0) = static_cast<value_type>(b(i0));
```

Comment:
afaikt our documentation does not state what users should expect here. With this we prefer assignable over the static cast ... which kind of changes the behavior. Some users might get hit by that if they only had `Views` of `Rank<3` and relied on the old behavior. I feel like this is quite a large change for 4.7.1 to fix stuff in Trilinos.